### ISSUE TITLE
How can I deserialize dynamic types?

### ISSUE THREAD

**Issue by elfchief** [go-restruct/restruct#19]
So I'm writing some code to deserialize a datastream where there are several hundred different types of packet, each with its own format, each with leading bytes to say what type the message is. I'm trying to unpack the packets by dynamically getting a struct of the appropriate type. Simplified, the code looks roughly like this:
```
type MessageStruct interface{}
type Messages map[uint8]MessageStruct
var MessageList = Messages{
    0x01: MessageTypeA
    0x02: MessageTypeB
}

msg_parsed := MessageList[message_id]
restruct.Unpack(msg, binary.LittleEndian, &msg_parsed)
```

...the problem being that restruct only sees the msg_parsed struct as type interface{}, so doesn't have anything to deserialize.

I realize that I could do a type switch:
```
switch msg_parsed.(type) {
case MessageTypeA:
    restruct.Unpack(msg. binary.LittleEndian, &msg_parsed.(MessageTypeA))
case MessageTypeB:
    restruct.Unpack(msg. binary.LittleEndian, &msg_parsed.(MessageTypeB))
}
```

... but that is very much not DRY, and gets unwieldy when you have a few hundred different possibilities.

Am I missing another way to accomplish this? Any suggestions?

**Comment by jchv** [comment]
There's a lot to be said about supporting prefix byte stuff for type switching, but of course a very general solution to that problem is not obvious.

As for the problem you're having, I'm pretty sure it's unrelated and has to do with interface boxing in Go. Just to be clear, the third argument to `restruct.Unpack` is, in fact, an `interface{}` - so whatever you pass will be interface-boxed. However, what you're doing is accidentally taking a _value type_ of `MessageTypeA` or `MessageTypeB`, boxing that into an `interface{}` type, then taking the address of that `interface{}` type, which then gets boxed into an `interface{}` again.

To give you an idea of what I mean, here's what I think is happening:

```go
&msg_parsed = &interface{}(MessageList[message_id])
&MessageList[message_id] = &MessageList[message_id]
```

You may, for example, want to try this:

```
type MessageStruct interface{}
type Messages map[uint8]MessageStruct
var MessageList = Messages{
    0x01: MessageTypeA{}
    0x02: MessageTypeB{}
}

msg_parsed := &MessageList[message_id]
restruct.Unpack(msg, binary.LittleEndian, msg_parsed)
```

If you do want to treat it as a value type, so that you don't modify the `MessageList` structure, you may need to play with the `reflect` library to effectively take a copy of the structure and make a pointer using it.

**Comment by elfchief** [comment]
> There's a lot to be said about supporting prefix byte stuff for type switching, but of course a very general solution to that problem is not obvious.

Yeah, I've been trying to think up a good general-purpose solution to this. It's actually worse for this particular chunk of decoding, because sometimes the exact structure to be decoded depends on either a) the length of the total message, or b) the value of a version value hidden at an arbitrary location inside each structure. Ugh. It would totally be nice to have a "if this field is X, use X1 structure, if this field is Y, use Y1 structure", though. (Also a "this byte should always be 0xNN, if it's not, fail unpacking)

> As for the problem you're having, I'm pretty sure it's unrelated and has to do with interface boxing in Go.

Ah, yes, good catch. I'm relatively new to golang, so the nuances here got a bit lost. Turning a bunch of stuff into pointers did indeed help out here. (It does have the problem that it updates the 'template' copy of the structure, of course. But yes, definitely happier:
```Decoding from:
([]uint8) (len=20 cap=3294) {
 00000000  80 2e 38 05 3f 7a fd ff  05 00 00 00 03 00 00 00  |..8.?z..........|
 00000010  72 00 00 00                                       |r...|
}
Decoding using:
(*main.UbxMsgNavClock)(0xc420014400)({
 ITOW: (uint32) 0,
 ClkB: (int32) 0,
 ClkD: (int32) 0,
 TAcc: (uint32) 0,
 FAcc: (uint32) 0
})
Decoded as:
(*main.UbxMsgNavClock)(0xc420014400)({
 ITOW: (uint32) 87568000,
 ClkB: (int32) -165313,
 ClkD: (int32) 5,
 TAcc: (uint32) 3,
 FAcc: (uint32) 114
})```

Using the suggested code you provided for a blank type, though, doesn't work, I'm guessing because it creates a nil object and restruct doesn't like that?

```Decoding using:
(**main.UbxMsgNavClock)(0xc42000c090)(<nil>)
Decoded as:
(**main.UbxMsgNavClock)(0xc42000c090)(<nil>)
```

I've tried a bunch of different permutations of this, and the shallow copying example you provided, and bunch of other ideas that came up when I was looking for more details on this type of thing, but I always end up with either the above, or with:

```Decoding using:
(reflect.Value) <**main.UbxMsgNavClock Value>
Decoded as:
(reflect.Value) <**main.UbxMsgNavClock Value>
```

I'm probably just missing something obvious, and I realize that at this point we're quite beyond anything that's actually a problem with restruct, but if you could point me the right direction here, it would very much be appreciated.

**Comment by jchv** [comment]
Not 100% sure, but I have some guesses. Firstly, when you have a `reflect.Value` you want to call `Value.Interface()` to turn it into an `interface{}`. This doesn't lose the type, and it's important because your reflect.Values most likely won't work for much directly. Secondly, it looks like you're creating a pointer to a pointer, i.e. `reflect.New(reflect.TypeOf(&main.UbxMsgNavClock{}))` - try instead `reflect.New(reflect.TypeOf(main.UbxMsgNavClock{}))` if so. `reflect.New` acts like Go's `new` operator, which will always allocate on heap and return a pointer of the type you ask for. So maybe something closer to this:

```go
value := reflect.New(reflect.TypeOf(main.UbxMsgNavClock{})).Interface()
restruct.Unpack(msg, binary.LittleEndian, value)
```

Let me know if that helps.

**Comment by elfchief** [comment]
So after way more pain and suffering than it was probably worth, the magic incantation ends up being:

```
msg_template := (*MessageList)[message_id]
val := reflect.Indirect(reflect.ValueOf(msg_template))
msg_parsed := reflect.New(val.Type()).Interface()
restruct.Unpack(msg, binary.LittleEndian, &msg_parsed)
```

...obviously restruct needs an "allocate me a new one of these" flag so it'll do that for me. ;)

Anyhow, yeah, looks good now, much appreciate the pointing in the right direction. Feel free to snag snippets from this as an example for the docs or such, if you'd like.  <3

**Comment by jchv** [comment]
Nice!

>...obviously restruct needs an "allocate me a new one of these" flag so it'll do that for me. ;)

That's not a terrible idea for a utility function to help with typeswitching needs. Realistically, since the uses for typeswitching in binary structures is so broad, it'd probably make sense to support not one, but multiple different approaches to it, so the most common approach can be easy but the most flexible approach can still be relatively verbose.

But I digress. I'll keep this case in mind for the future.
